api/app/Models/PoolCandidate.php

```diff
@@ -834,4 +839,112 @@ public function setApplicationSnapshot()
         $this->profile_snapshot = $profile;
         $this->save();
     }
+
+    public function computeDecisionsPerAssessmentStep()
+    {
+        $decisions = [];
+        $this->load([
+            'pool.assessmentSteps',
+            'assessmentResults',
+        ]);
+
+        foreach ($this->pool->assessmentSteps as $step) {
+            $stepId = $step->id;
+            $hasFailure = false;
+            $hasOnHold = false;
+            $hasToAssess = false;
+
+            $stepResults = $this->assessmentResults()->where('assessment_step_id', $stepId);
+            if (! $stepResults->count()) {
+                $decisions[$stepId] = null;
+
+                continue;
+            }
+
+            // Check assessed essential skills on this step
+            $essentialSkillAssessments = $stepResults->whereHas('assessmentStep', function ($query) {
+                $query->whereHas('poolSkills', function ($query) {
+                    $query->where('type', PoolSkillType::ESSENTIAL->name);
+                });
+            });
+
+            if (! $essentialSkillAssessments) {
```

Comment:
What if an assessment step doesn't _have_ any required skills?

Also, remember that the initial step (APPLICATION_SCREENING) also requires an education-type assessment result. 
EDIT: nevermind, I see you handle that later 😆

Reply:
oph right, we changed it so that we can do asset only right? So instead of setting and continue I should be doing `$hasToAssess = true;` instead, right?

Reply:
Actually... if a step has no essential skills, and its not APPLICATION ASSESSMENT (ie requires Education), then... is it automatically successful? If that's the case, you can do nothing here, and the function will return SUCCESSFUL correctly.

Reply:
So just to confirm, you are suggesting we just delete this check all together and it should function as expected?

Reply:
Yes.

Reply:
This is exactly the kind of situation which should have a phpunit test

Reply:
Cool, thanks! Yeah, my next task is writing tests to cover different scenarios now that I think the logic is worked out :+1:

Reply:
Alright, tests added. I think I got all the logic covered but let me know if I am missing anything!

[add tests for logic](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10642/commits/138c4be4cea3ba71a471e6e7938b98a45d8ffc1c)

Reply:
I may need some assistance. The essential skill test is flaky and I really cannot tell why. I am guessing something to do with the seeded data but have not be able to locate it :confounded: 

https://github.com/GCTC-NTGC/gc-digital-talent/actions/runs/9504282927/job/26196583002?pr=10642#step:7:561

Reply:
Okay, I think I got it fixed in https://github.com/GCTC-NTGC/gc-digital-talent/pull/10642/commits/fc2ab6bbbe70f4bc9d71916f81f9ce3eb0d811f1
All successes on executing 
`for i in {1..20}; do docker-compose exec -w /home/site/wwwroot/api webserver sh -c "./vendor/bin/phpunit ./tests/Feature/CandidateAssessmentStatusTest.php"; done`

Issue being, the test skill would vary between technical and behavioural. Assessment step one I believe is always application screening, which ignores behavioural skills. So any time a behavioural skill was created, the screening step ignored it.

Reply:
Thank-you! That makes sense now :pray: